File: sdk/flybody.mjs
/** Portable browser / Node client for FLYLAB Body API v1. No dependencies. */
export class BodyAPIError extends Error {
  constructor(status, response) {
    super(`Body API ${status}: ${JSON.stringify(response)}`);
    this.status = status;
    this.response = response;
  }
}
export class FlyBody {
  constructor(baseUrl = 'http://127.0.0.1:8787') {
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.sessionId = null;
    this.episodeId = null;
    this.stepIndex = 0;
    this.inFlight = false;
  }
  async post(path, payload) {
    if (this.inFlight) throw new Error('Await the previous body request first');
    this.inFlight = true;
    try {
      const headers = {'Content-Type': 'application/json', 'X-Flybrain-Local': '1'};
      // Browsers supply their own Origin; Node needs this local-origin marker.
      if (typeof window === 'undefined') headers.Origin = this.baseUrl;
      const body = payload === undefined ? undefined : JSON.stringify(payload);
      let response;
      for (let attempt = 0; attempt < 2; attempt++) {
        try {
          response = await fetch(`${this.baseUrl}/api/v1/body${path}`, {
            method: 'POST', headers, body, signal: AbortSignal.timeout(30000),
          });
          break;
        } catch (error) { if (attempt) throw error; }
      }
      const result = await response.json();
      if (!response.ok) throw new BodyAPIError(response.status, result);
      return result;
    } finally { this.inFlight = false; }
  }
  path(command) {
    if (!this.sessionId) throw new Error('Call connect() first');
    return `/sessions/${this.sessionId}/${command}`;
  }
  async connect({controller_name = 'web-game', preset = 'descending', ...config} = {}) {
    const result = await this.post('/sessions', {
      request_id: crypto.randomUUID(), controller_name, preset, ...config,
    });
    this.sessionId = result.session_id;
    this.episodeId = result.episode_id;
    this.stepIndex = result.step_index;
    return result;
  }
  async reset({seed = 0, reset_learning = true} = {}) {
    const result = await this.post(this.path('reset'), {
      request_id: crypto.randomUUID(), seed, reset_learning,
    });
    this.episodeId = result.episode_id;
    this.stepIndex = 0;
    return result;
  }
  async step(observation, {dt_ms = 20} = {}) {
    if (!this.episodeId) throw new Error('Call reset() before step()');
    const result = await this.post(this.path('step'), {
      episode_id: this.episodeId, step_index: this.stepIndex + 1, dt_ms, observation,
    });
    this.stepIndex = result.step_index;
    return result;
  }
  async reward(value, {components = {}, terminated = false, truncated = false} = {}) {
    return this.post(this.path('reward'), {
      episode_id: this.episodeId, step_index: this.stepIndex, value, components,
      terminated, truncated,
    });
  }
  async release() {
    if (!this.sessionId) return;
    const result = await this.post(this.path('release'));
    this.sessionId = this.episodeId = null;
    return result;
  }
}

/** Convert a small RGB(A) render target into the API's normalized luminance frame.
 * Use flipY:true for raw WebGL readPixels (whose origin is bottom-left).
 * Alpha is ignored; sRGB is linearized before luminance conversion.
 */
export function visionFromRGBA(rgba, width, height, {flipY = false} = {}) {
  if (width < 1 || height < 1 || width > 128 || height > 128 ||
      !Number.isInteger(width) || !Number.isInteger(height) || rgba.length !== width * height * 4)
    throw new Error('Expected a 1–128 px RGBA frame on each axis');
  const pixels = new Array(width * height);
  const linear = value => {const v = value / 255; return v <= .04045 ? v / 12.92 : ((v + .055) / 1.055) ** 2.4};
  for (let y = 0; y < height; y++) for (let x = 0; x < width; x++) {
    const i = ((flipY ? height - 1 - y : y) * width + x) * 4;
    pixels[y * width + x] = linear(rgba[i]) * .2126 + linear(rgba[i + 1]) * .7152 + linear(rgba[i + 2]) * .0722;
  }
  return {width, height, pixels};
}

File: sdk/flybody.py
"""Small synchronous Python client for the local FLYLAB Body API v1."""

from __future__ import annotations

from typing import Any, Self
from uuid import uuid4

import httpx


class BodyAPIError(RuntimeError):
    """A protocol rejection; status and response retain the server's machine-readable detail."""

    def __init__(self, status: int, response: Any) -> None:
        self.status = status
        self.response = response
        super().__init__(f"Body API {status}: {response}")


class FlyBody:
    """One controller per brain; call step, advance your world, then submit its reward.

    Transport timeouts retry the exact request once. No request changes neural time
    twice when repeated at the same API boundary. Use one sequential control loop.
    """

    def __init__(
        self,
        base_url: str = "http://127.0.0.1:8787",
        *,
        controller_name: str = "python-game",
        preset: str = "descending",
        timeout: float = 30,
        **configuration: Any,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.client = httpx.Client(
            base_url=self.base_url,
            timeout=timeout,
            trust_env=False,
            headers={"Origin": self.base_url, "X-Flybrain-Local": "1"},
        )
        self.config = {
            "request_id": str(uuid4()),
            "controller_name": controller_name,
            "preset": preset,
            **configuration,
        }
        self.session_id: str | None = None
        self.episode_id: str | None = None
        self.step_index = 0

    def _post(self, path: str, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        for attempt in range(2):
            try:
                response = self.client.post("/api/v1/body" + path, json=payload)
                break
            except httpx.TransportError:
                if attempt:
                    raise
        result: dict[str, Any] = response.json()
        if not response.is_success:
            raise BodyAPIError(response.status_code, result)
        return result

    def _path(self, command: str) -> str:
        if self.session_id is None:
            raise RuntimeError("Call connect() first")
        return f"/sessions/{self.session_id}/{command}"

    def connect(self) -> dict[str, Any]:
        result = self._post("/sessions", self.config)
        self.session_id = result["session_id"]
        self.episode_id = result["episode_id"]
        self.step_index = result["step_index"]
        return result

    def reset(self, seed: int = 0, *, reset_learning: bool = True) -> dict[str, Any]:
        result = self._post(
            self._path("reset"),
            {"request_id": str(uuid4()), "seed": seed, "reset_learning": reset_learning},
        )
        self.episode_id, self.step_index = result["episode_id"], 0
        return result

    def step(self, observation: dict[str, Any], dt_ms: int = 20) -> dict[str, Any]:
        if self.episode_id is None:
            raise RuntimeError("Call reset() before step()")
        result = self._post(
            self._path("step"),
            {
                "episode_id": self.episode_id,
                "step_index": self.step_index + 1,
                "dt_ms": dt_ms,
                "observation": observation,
            },
        )
        self.step_index = result["step_index"]
        return result

    def reward(
        self,
        value: float,
        *,
        components: dict[str, float] | None = None,
        terminated: bool = False,
        truncated: bool = False,
    ) -> dict[str, Any]:
        return self._post(
            self._path("reward"),
            {
                "episode_id": self.episode_id,
                "step_index": self.step_index,
                "value": value,
                "components": components or {},
                "terminated": terminated,
                "truncated": truncated,
            },
        )

    def release(self) -> dict[str, Any] | None:
        if self.session_id is None:
            return None
        result = self._post(self._path("release"))
        self.session_id = self.episode_id = None
        return result

    def close(self) -> None:
        try:
            self.release()
        finally:
            self.client.close()

    def __enter__(self) -> Self:
        self.connect()
        return self

    def __exit__(self, *args: object) -> None:
        self.close()
